### DISCUSSION TITLE
Set and use custom variables in Markdown files

### DISCUSSION THREAD

**Discussion by TheEvilSkeleton** [getpelican/pelican#3343] [Q&A] [answered]
I'm trying to set variables within Markdown files so I can reuse them everywhere else (the Markdown files themselves and HTML templates).

Say I write an article in Markdown where the file name is `keyboard-review.md`, which contains:

```markdown
Title: My First Review
Date: 2010-12-03 10:20
Category: Review
Description: Following is a review of my favorite mechanical keyboard.

{{ Description }}
```

Ideally, `{{ Description }}` is supposed to be a variable that I can reuse everywhere within the Markdown file, but it ends up printing nothing.

From what I understand, I can use the `{% set %}` statement/function to set and use variables within the same Markdown file, i.e. `{% set Description = [...] %}`, but this method has a limitation with the scope, where I can't access them in HTML templates. In other words, one method only works in Markdown files, and the other method only works in HTML templates. I can't find a method that has _both_.

Would anybody be able to help me out on this? Maybe I missed something in the docs...

Edit: Also opened https://github.com/pelican-plugins/jinja2content/issues/8 as I'm not sure where to open it.

**Comment by justinmayer** [comment]
Hi Hari. Are you _absolutely certain_ you reviewed the documentation before asking your question, including the [Frequently Asked Questions](https://docs.getpelican.com/en/latest/faq.html#can-i-use-arbitrary-metadata-in-my-templates)?

As far as I am aware, Pelican itself only supports variables in _templates_ and **not** inside _Markdown content_. You might find the latter functionality in the form of a plugin, however.

The above-linked documentation describes how to handle this behavior in templates.

**Reply by TheEvilSkeleton** [reply]
Yes, I have, but I forgot to link this page.

From what I understood and tested, these variables can only be accessed in templates, but not in Markdown files. If I apply that example in the Markdown file and try to access that variable in the Markdown file, I'll get an error.

For example, `keyboard-review.md` contains the following:
```
Title: My First Review
Date: 2010-12-03 10:20
Category: Review
Modified: 2012-08-08

{% if article.modified %}
Last modified: {{ article.modified }}
{% endif %}
```

but I get this error:
```
/var/home/TheEvilSkeleton/Projects/theevilskeleton.gitlab.io.pelican/content/articles/keyboard-review.md. re-generating...
[13:25:43] ERROR    Could not process articles/keyboard-review.md                                                                              log.py:94
                    'article' is undefined                                                                                                              
```

I also forgot to mention that I'm using [jinja2content](https://github.com/pelican-plugins/jinja2content) to use Jinja2 code in Markdown files.

**Reply by justinmayer** [reply, marked as answer]
I updated my answer above to reflect that Pelican by itself does not process Jinja directives in Markdown content. How the Jinja2Content plugin works is outside the scope of Pelican core.

**Reply by TheEvilSkeleton** [reply]
Right, fair. I edited my original post yesterday and included https://github.com/pelican-plugins/jinja2content/issues/8. I'll close this issue and hope to have my question answered there. Thanks!
